### PR TITLE
Introduce local block factory

### PR DESCRIPTION
**Pull request by dnhatn**
Requesting and returning memory from a CircuitBreaker can be costly due to the involvement of read/write on one or several atomic longs. To address this issue, the local breaker adopts a strategy of over-requesting memory, utilizing the reserved amount for subsequent memory requests without direct access to the actual breaker.

Before passing a Block to another Driver, it is necessary to switch the owning block factory to its parent, which is associated with the global breaker. This is done to bypass the local breaker when releasing memory, as the releasing thread can be any thread, not necessarily the one executing the Driver.

There are two specific operators that need to change the owning block factory: SinkOperator (superset of ExchangeSinkOperator), which is the last operator of a Driver, and AsyncOperator, which can be responded by any thread in response.

The optimization reduces the latency of the enrich operation in the nyc_taxis benchmark from 100ms to 50ms. When combined with #102902, it further reduces the latency to below 40ms, better than the previous performance before the regression.

Relates #102625

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java [hunk 1]
```diff
@@ -11,13 +11,19 @@
...
 import org.elasticsearch.common.util.BytesRefArray;
 import org.elasticsearch.compute.data.Block.MvOrdering;
 
 import java.util.BitSet;
 
 public class BlockFactory {
+    public static final String LOCAL_BREAKER_OVER_RESERVED_SIZE_SETTING = "esql.block_factory.local_breaker.over_reserved";
+    public static final ByteSizeValue LOCAL_BREAKER_OVER_RESERVED_DEFAULT_SIZE = ByteSizeValue.ofKb(4);
+
+    public static final String LOCAL_BREAKER_OVER_RESERVED_MAX_SIZE_SETTING = "esql.block_factory.local_breaker.max_over_reserved";
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:25]: These are internal settings now; we can promote them later.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/LocalCircuitBreaker.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+/**
+ * Requesting and returning memory from a {@link CircuitBreaker} can be costly due to the involvement of read/write
+ * on one or several atomic longs. To address this issue, the local breaker adopts a strategy of over-requesting memory,
+ * utilizing the reserved amount for subsequent memory requests without direct access to the actual breaker.
+ *
+ * @see BlockFactory#newChildFactory(LocalCircuitBreaker)
+ * @see Block#allowPassingToDifferentDriver()
+ */
+public final class LocalCircuitBreaker implements CircuitBreaker, Releasable {
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/LocalCircuitBreaker.java:25]: ++

**Review by ChrisHegarty — COMMENTED** [review]
The approach is very nice 👍 . Layering things through a simple hierarchy in block factory, and switching back to the parent before transfer is clever - it allows pages and blocks transferred out seamlessly "give back".

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/SinkOperator.java [hunk 3]
```diff
@@ -23,6 +23,15 @@ public final Page getOutput() {
         throw new UnsupportedOperationException();
     }
 
+    protected abstract void doAddInput(Page page);
+
+    @Override
+    public final void addInput(Page page) {
+        // We need to change the ownership of the blocks of the input page before passing them to another driver.
+        page.allowPassingToDifferentDriver();
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/SinkOperator.java:31]: ++ yes, it's important to transfer ownership in the executing thread *before* the page leaves.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java [hunk 4]
```diff
@@ -102,9 +102,15 @@ public interface Block extends Accountable, BlockLoader.Block, NamedWriteable, R
...
      * @param positions the positions to retain
      * @return a filtered block
+     * TODO: pass BlockFactory
      */
     Block filter(int... positions);
 
+    /**
+     * This method must be called before passing this Block to another Driver.
+     */
+    void allowPassingToDifferentDriver();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java]: ++, I've pushed https://github.com/elastic/elasticsearch/pull/102901/commits/341dd8d42fd98cce1bb6c870d43c0cd2b055fd3f

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thank you, Chris!

**Pull request merged**